File: apps/mobile-android/app/src/main/kotlin/br/com/tatame/core/designsystem/components/BeltBar.kt
package br.com.tatame.core.designsystem.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.tatame.designsystem.tokens.LumiraTokens

/**
 * `BeltDef`-shaped belt payload the design system renders (GRD.18). Mirrors
 * the API's `currentBelt` contract — components are keyed by data (slugs),
 * never by hardcoded belt names (resolved ticket design-system/04).
 */
data class BeltDisplay(
    val colorSlug: String, // "belt.blue" (or bare "blue")
    val tipColorSlug: String? = null, // null = default belt.tip; black belt sends "belt.red"
    val degrees: Int,
    val maxDegrees: Int, // 0 = no degree stripes (red belt in v1)
)

/** BeltBar sizes per the resolved anatomy: sm list rows, md cards, lg hero. */
enum class BeltBarSize(val height: Dp, val stripeWidth: Dp) {
    Sm(10.dp, 2.dp),
    Md(14.dp, 3.dp),
    Lg(20.dp, 4.dp),
}

/**
 * Pure layout/color rules of the BeltBar anatomy — JVM-unit-testable, no
 * Compose runtime (resolved ticket design-system/04 `## Answer` §4).
 */
object BeltBarLayout {

    /** Ponteira occupies ~22% of the bar width. */
    const val TIP_WIDTH_FRACTION = 0.22f

    /** Static, brand-independent belt color map — exempt from white-label AND dark remix. */
    val BELT_COLORS: Map<String, Color> = mapOf(
        "white" to LumiraTokens.Colors.BeltWhite,
        "gray" to LumiraTokens.Colors.BeltGray,
        "yellow" to LumiraTokens.Colors.BeltYellow,
        "orange" to LumiraTokens.Colors.BeltOrange,
        "green" to LumiraTokens.Colors.BeltGreen,
        "blue" to LumiraTokens.Colors.BeltBlue,
        "purple" to LumiraTokens.Colors.BeltPurple,
        "brown" to LumiraTokens.Colors.BeltBrown,
        "black" to LumiraTokens.Colors.BeltBlack,
        "red" to LumiraTokens.Colors.BeltRed,
    )

    /** Accepts both token-slug (`belt.blue`) and bare (`blue`) spellings. */
    fun normalizeSlug(slug: String): String = slug.removePrefix("belt.")

    /** null = unknown slug (caller falls back to gray + logs a warning). */
    fun barColor(colorSlug: String): Color? = BELT_COLORS[normalizeSlug(colorSlug)]

    /** Defensive gray fallback keeps old clients alive if the catalog grows first. */
    fun barColorOrFallback(colorSlug: String): Color =
        barColor(colorSlug) ?: LumiraTokens.Colors.BeltGray

    /** `tipColorSlug ?? belt.tip` — black belt sends `belt.red` (IBJJF red ponteira). */
    fun tipColor(tipColorSlug: String?): Color =
        tipColorSlug?.let { BELT_COLORS[normalizeSlug(it)] } ?: LumiraTokens.Colors.BeltTip

    /**
     * White degree stripes drawn ON the ponteira: clamped to `maxDegrees`;
     * `maxDegrees == 0` (red belt in v1) renders no stripes at all.
     */
    fun stripeCount(degrees: Int, maxDegrees: Int): Int =
        if (maxDegrees <= 0) 0 else degrees.coerceIn(0, maxDegrees)
}

/**
 * The belt drawn with primitives per the resolved anatomy (GRD.18): rounded
 * bar (radius-xs) filled by the belt color, inset `belt.outline` hairline
 * (what keeps Branca visible — never "fixed" with a gray fill), ~22% ponteira
 * in `tipColorSlug ?? belt.tip`, white degree stripes on the ponteira. Black
 * belt renders white dan stripes on the red tip; red belt renders no stripes;
 * unknown `colorSlug` falls back to gray with a logged warning. Width is
 * fluid (caller's modifier); height fixed per [BeltBarSize].
 */
@Composable
fun BeltBar(
    belt: BeltDisplay,
    modifier: Modifier = Modifier,
    size: BeltBarSize = BeltBarSize.Md,
) {
    val barColor = BeltBarLayout.barColor(belt.colorSlug)
    remember(belt.colorSlug) {
        if (barColor == null) {
            android.util.Log.w(
                "BeltBar",
                "Unknown belt colorSlug '${belt.colorSlug}' — falling back to gray",
            )
        }
        belt.colorSlug
    }
    val shape = RoundedCornerShape(LumiraTokens.Radius.Xs)
    Box(
        modifier = modifier
            .height(size.height)
            .clip(shape)
            .background(barColor ?: LumiraTokens.Colors.BeltGray),
    ) {
        // Ponteira with the degree stripes.
        Row(
            modifier = Modifier
                .align(Alignment.CenterEnd)
                .fillMaxHeight()
                .fillMaxWidth(BeltBarLayout.TIP_WIDTH_FRACTION)
                .background(BeltBarLayout.tipColor(belt.tipColorSlug)),
            horizontalArrangement = Arrangement.SpaceEvenly,
        ) {
            repeat(BeltBarLayout.stripeCount(belt.degrees, belt.maxDegrees)) {
                Box(
                    modifier = Modifier
                        .width(size.stripeWidth)
                        .fillMaxHeight()
                        .background(LumiraTokens.Colors.BeltStripe),
                )
            }
        }
        // Inset hairline outline over bar + ponteira (uniformly applied).
        Box(Modifier.matchParentSize().border(1.dp, LumiraTokens.Colors.BeltOutline, shape))
    }
}

/**
 * Chip variant (GRD.18): mini belt swatch + PT-BR label in a pill — profile
 * belt chips ("Faixa preta · 2º dan") and Graduações válidas régua chips.
 * [enabled] = false dims the chip (kids belt toggled off by the admin).
 */
@Composable
fun BeltChip(
    label: String,
    belt: BeltDisplay,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
    Row(
        modifier = modifier
            .alpha(if (enabled) 1f else 0.45f)
            .background(
                color = LumiraTokens.Colors.BgSurface,
                shape = RoundedCornerShape(LumiraTokens.Radius.Pill),
            )
            .border(
                width = 1.dp,
                color = LumiraTokens.Colors.Border1,
                shape = RoundedCornerShape(LumiraTokens.Radius.Pill),
            )
            .padding(horizontal = LumiraTokens.Space.S3, vertical = LumiraTokens.Space.S1),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(LumiraTokens.Space.S2),
    ) {
        BeltBar(belt = belt, size = BeltBarSize.Sm, modifier = Modifier.width(28.dp))
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = LumiraTokens.Colors.Fg2,
        )
    }
}

File: apps/mobile-ios/Packages/DesignSystem/Sources/DesignSystem/Components/BeltBar.swift
// BeltBar — the one belt-rendering component (spec 005, GRD.21; resolved
// belt-tokenization anatomy). Drawn with SwiftUI primitives, never images:
// rounded bar filled by the colorSlug token, inset `belt.outline` hairline
// (what keeps Branca visible — never "fixed" with gray), ponteira ~22% width
// in `tipColorSlug ?? belt.tip`, white degree stripes on the ponteira.
// Keyed by data (`BeltDef`-shaped payloads) — zero hardcoded ladders; an
// unknown colorSlug falls back to gray with a logged warning.

import os
import SwiftUI

/// Pure rendering rules behind `BeltBar` (unit-tested view-model helpers).
public enum BeltBarSpec {
    /// `color.belt.*` slug → generated Lumira token. Belts are static,
    /// brand-independent tokens — exempt from white-label and dark remix.
    public static let colorsBySlug: [String: Color] = [
        "belt.white": LumiraTokens.Colors.beltWhite,
        "belt.gray": LumiraTokens.Colors.beltGray,
        "belt.yellow": LumiraTokens.Colors.beltYellow,
        "belt.orange": LumiraTokens.Colors.beltOrange,
        "belt.green": LumiraTokens.Colors.beltGreen,
        "belt.blue": LumiraTokens.Colors.beltBlue,
        "belt.purple": LumiraTokens.Colors.beltPurple,
        "belt.brown": LumiraTokens.Colors.beltBrown,
        "belt.black": LumiraTokens.Colors.beltBlack,
        "belt.red": LumiraTokens.Colors.beltRed,
    ]

    /// True when the slug maps to a shipped belt token.
    public static func isKnown(colorSlug: String) -> Bool {
        colorsBySlug[colorSlug] != nil
    }

    /// Bar fill for a slug — unknown slugs render gray (defensive default;
    /// keeps old clients alive if the catalog gains a color first) and log
    /// a warning.
    public static func fill(colorSlug: String) -> Color {
        if let color = colorsBySlug[colorSlug] {
            return color
        }
        logger.warning("Unknown belt colorSlug '\(colorSlug, privacy: .public)' — rendering gray fallback")
        return LumiraTokens.Colors.beltGray
    }

    /// Ponteira fill: `tipColorSlug ?? belt.tip` (black belt sends belt.red).
    public static func tipFill(tipColorSlug: String?) -> Color {
        guard let tipColorSlug else { return LumiraTokens.Colors.beltTip }
        return fill(colorSlug: tipColorSlug)
    }

    /// Degree stripes drawn on the ponteira: current degrees clamped into
    /// 0...maxDegrees; a zero maxDegrees (red belt in v1) never stripes.
    public static func stripeCount(degrees: Int, maxDegrees: Int) -> Int {
        guard maxDegrees > 0 else { return 0 }
        return min(max(degrees, 0), maxDegrees)
    }

    /// Ponteira share of the bar width (anatomy: ~22%).
    public static let tipFraction: CGFloat = 0.22

    private static let logger = Logger(subsystem: "app.tatame.ios", category: "BeltBar")
}

/// The drawn belt (bar + ponteira + degree stripes), sizes sm/md/lg.
public struct BeltBar: View {
    public enum Size: Sendable {
        /// List rows (roll call, dependent cards).
        case sm
        /// Cards (aluno home, timeline markers, profile).
        case md
        /// Hero (Graduação screen).
        case lg

        var height: CGFloat {
            switch self {
            case .sm: 10
            case .md: 16
            case .lg: 22
            }
        }

        var stripeWidth: CGFloat {
            switch self {
            case .sm: 2
            case .md: 3
            case .lg: 4
            }
        }
    }

    private let colorSlug: String
    private let tipColorSlug: String?
    private let degrees: Int
    private let maxDegrees: Int
    private let size: Size

    public init(colorSlug: String, tipColorSlug: String?, degrees: Int, maxDegrees: Int, size: Size = .md) {
        self.colorSlug = colorSlug
        self.tipColorSlug = tipColorSlug
        self.degrees = degrees
        self.maxDegrees = maxDegrees
        self.size = size
    }

    public var body: some View {
        GeometryReader { proxy in
            let radius = min(LumiraTokens.Radius.xs, size.height / 2)
            let tipWidth = proxy.size.width * BeltBarSpec.tipFraction
            ZStack(alignment: .trailing) {
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .fill(BeltBarSpec.fill(colorSlug: colorSlug))
                // Ponteira: solid block at the right end (~22% width).
                BeltTipShape(radius: radius)
                    .fill(BeltBarSpec.tipFill(tipColorSlug: tipColorSlug))
                    .frame(width: tipWidth)
                // Degree stripes: white, evenly spaced on the ponteira.
                let stripes = BeltBarSpec.stripeCount(degrees: degrees, maxDegrees: maxDegrees)
                if stripes > 0 {
                    HStack(spacing: 0) {
                        ForEach(0..<stripes, id: \.self) { _ in
                            Rectangle()
                                .fill(LumiraTokens.Colors.beltStripe)
                                .frame(width: size.stripeWidth)
                                .frame(maxWidth: .infinity)
                        }
                    }
                    .frame(width: tipWidth)
                }
                // Inset hairline outline — uniform on every belt; the spec
                // exists so Branca holds its edge on any surface.
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .strokeBorder(LumiraTokens.Colors.beltOutline, lineWidth: 1)
                    .frame(width: proxy.size.width)
            }
        }
        .frame(height: size.height)
        .accessibilityHidden(true)
    }
}

/// Right-end tip block that follows the bar's rounded trailing corners.
private struct BeltTipShape: Shape {
    let radius: CGFloat

    func path(in rect: CGRect) -> Path {
        Path(
            roundedRect: rect,
            cornerRadii: RectangleCornerRadii(
                topLeading: 0,
                bottomLeading: 0,
                bottomTrailing: radius,
                topTrailing: radius
            ),
            style: .continuous
        )
    }
}

/// Belt chip variant: a mini drawn belt + PT-BR label in a pill (profile
/// headers, graduações válidas, list rows).
public struct BeltChip: View {
    public enum Style: Sendable {
        /// Surface pill (lists, régua chips).
        case neutral
        /// Ink pill (profile header chips — professor-12 / aluno-19).
        case prominent
        /// Dimmed régua chip (kids belt disabled by the admin).
        case disabled
    }

    private let label: String
    private let colorSlug: String
    private let tipColorSlug: String?
    private let degrees: Int
    private let maxDegrees: Int
    private let style: Style

    public init(
        label: String,
        colorSlug: String,
        tipColorSlug: String? = nil,
        degrees: Int = 0,
        maxDegrees: Int = 0,
        style: Style = .neutral
    ) {
        self.label = label
        self.colorSlug = colorSlug
        self.tipColorSlug = tipColorSlug
        self.degrees = degrees
        self.maxDegrees = maxDegrees
        self.style = style
    }

    public var body: some View {
        HStack(spacing: LumiraTokens.Space.s2) {
            BeltBar(
                colorSlug: colorSlug,
                tipColorSlug: tipColorSlug,
                degrees: degrees,
                maxDegrees: maxDegrees,
                size: .sm
            )
            .frame(width: 34)
            Text(label)
                .font(.system(size: LumiraTokens.FontSize.textXs, weight: .semibold, design: .rounded))
                .foregroundStyle(foreground)
        }
        .padding(.horizontal, LumiraTokens.Space.s3)
        .padding(.vertical, LumiraTokens.Space.s1)
        .background(background)
        .clipShape(Capsule())
        .overlay(
            Capsule().strokeBorder(border, lineWidth: style == .prominent ? 0 : 1)
        )
        .opacity(style == .disabled ? 0.45 : 1)
    }

    private var foreground: Color {
        switch style {
        case .prominent: LumiraTokens.Colors.fgOnColor
        case .neutral, .disabled: LumiraTokens.Colors.fg2
        }
    }

    private var background: Color {
        switch style {
        case .prominent: LumiraTokens.Colors.fg1
        case .neutral, .disabled: LumiraTokens.Colors.bgSurface
        }
    }

    private var border: Color {
        switch style {
        case .prominent: .clear
        case .neutral, .disabled: LumiraTokens.Colors.border1
        }
    }
}
